plugins/docker_binary_builder/app/models/binary_builder.rb

```diff
@@ -0,0 +1,136 @@
+class BinaryBuilder
+  DOCKER_BUILD_FILE = 'Dockerfile.build'.freeze
+  BUILD_SCRIPT = '/app/build.sh'.freeze
+  ARTIFACTS_FILE = 'artifacts.tar'.freeze
+  ARTIFACTS_FILE_PATH = "/app/#{ARTIFACTS_FILE}".freeze
+  DOCKER_HOST_CACHE_DIR = '/opt/samson_build_cache'.freeze
+  CONTAINER_CACHE_DIR = '/build/cache'.freeze
+
+  def initialize(dir, project, reference, output)
+    @dir = dir
+    @project = project
+    @git_reference = reference
+    @output_stream = output
+  end
+
+  def build
+    return unless @project.try(:deploy_with_docker?) && File.exists?(File.join(@dir, DOCKER_BUILD_FILE))
+    @output_stream.puts "Connecting to Docker host with Api version: #{docker_api_version} ..."
+
+    @image = create_build_image
+    @container = Docker::Container.create(create_container_options)
+
+    start_build_script
+    retrieve_binaries
+
+    @output_stream.puts 'Continuing docker build...'
+  ensure
+    @output_stream.puts 'Cleaning up docker build image and container...'
+    @container.delete(force: true) if @container
+    @image.remove(force: true) if @image
+  end
+
+  private
+
+  def start_build_script
+    @output_stream.puts 'Now starting Build container...'
+    @container.tap(&:start).attach { |_stream, chunk| @output_stream.write chunk }
+  rescue => ex
+    @output_stream.puts "Failed to run the build script '#{BUILD_SCRIPT}' inside container."
+    raise ex
+  end
+
+  def retrieve_binaries
+    @output_stream.puts "Grabbing '#{ARTIFACTS_FILE_PATH}' from build container..."
+    artifacts_tar = Tempfile.new(['artifacts', '.tar'], @dir)
+    artifacts_tar.binmode
+    @container.copy(ARTIFACTS_FILE_PATH) { |chunk| artifacts_tar.write chunk }
+    artifacts_tar.close
+
+    untar(artifacts_tar.path)
```

Comment:
Given a deterministic build process with multiple artifacts, since the mtimes are different after un-tar, docker wouldn't be able to cache steps. Really not a big deal, probably nothing in practice.